subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -181,7 +181,8 @@ public void validateMutation(MutationType type) {
     private boolean transitive = true;
     private Set<Configuration> extendsFrom = new LinkedHashSet<Configuration>();
     private String description;
-    private Set<ExcludeRule> excludeRules = new LinkedHashSet<ExcludeRule>();
+    private final Set<Object> excludeRules = new LinkedHashSet<Object>();
+    private Set<ExcludeRule> parsedExcludeRules;
```

Comment:
Some food for thought: I'm wondering whether we could use a `SetProperty` or maybe a `LockableSetProperty` to implement this pattern. The former would compute things every time which is not what we want. The latter provides the laziness with one time evaluation of passed `Providers` but does not allow later modification so it wouldn't work out of the box either. But it does sound like a potential use case for the Provider API.

Reply:
Interesting idea! Another approach would be making the parsers much more efficient. Right now the MapNotationParser used here is just very slow. If we could fix that there would be no need for laziness.